### PR TITLE
Add an LRU cache to make_trie_root_and_nodes()

### PR DESCRIPTION
**Pull request by gsalgado**

### CODE REVIEW

**Comment by gsalgado** [comment]
Unfortunately this won't be enough by itself because we'd still be making one `block.get_receipts()` call for every tx applied, and that's too expensive to run for every tx.

**Comment by gsalgado** [comment]
And caching `ChainDB.get_receipts()` doesn't help because every time it is called (as part of `VM.apply_transaction()`) it is passed a different `receipt_root` for the trie containing only the receipts of the transactions that have already been applied

**Comment by gsalgado** [comment]
So, this won't replace #565 but it might still be beneficial, specially since Byzantium blocks may have duplicate receipt_roots. Will benchmark to find out

**Comment by gsalgado** [comment]
Now that #565 has been merged and we're no longer generating/storing partial tx/receipt tries, it makes more sense to cache the whole thing here.

File: evm/db/trie.py [hunk 1]
```diff
@@ -1,19 +1,33 @@
...
-        index_key = rlp.encode(index, sedes=rlp.sedes.big_endian_int)
-        transaction_db[index_key] = rlp.encode(transaction)
 
-    return transaction_db.root_hash, transaction_db.db.wrapped_db.kv_store
+# This cache is expected to be useful when importing blocks as we call this once when importing
+# and again when validating the imported block. But it should also help for post-Byzantium blocks
+# as it's common for them to have duplicate receipt_roots. Given that, it probably makes sense to
+# use a relatively small cache size here.
+@functools.lru_cache(128)
+def _make_trie_root_and_nodes(items: Tuple[bytes, ...]) -> Tuple[bytes, Dict]:
```
Comment (pipermerriam) [inline evm/db/trie.py]: Return type should be `Tuple[bytes, Dict[bytes, bytes]]` right?

Comment (gsalgado) [reply evm/db/trie.py]: Indeed. I guess mypy considers a bare `Dict` to be `Dict[Any, Any]`

**Pull request merged**